examples/experimental/trial/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,172 @@
+        self.logger = TorchWriter()
+
+        # Create a unique download directory for each rank so they don't overwrite each other.
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_downloaded = False
+
+        mnist_shape = (1, 28, 28)
+        self.generator =  self.context._Model(Generator(latent_dim=self.context.get_hparam("latent_dim"), img_shape=mnist_shape))
+        self.discriminator =  self.context._Model(Discriminator(img_shape=mnist_shape))
+
+        lr = self.context.get_hparam("lr")
+        b1 = self.context.get_hparam("b1")
+        b2 = self.context.get_hparam("b2")
+
+        self.opt_g = self.context._Optimizer(torch.optim.Adam(self.generator.parameters(), lr=lr, betas=(b1, b2)))
+        self.opt_d = self.context._Optimizer(torch.optim.Adam(self.discriminator.parameters(), lr=lr, betas=(b1, b2)))
+
+        self.lr_g = self.context._LRScheduler(
+            lr_scheduler=LambdaLR(self.opt_g, lr_lambda=lambda epoch: 0.95 ** epoch),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+    def build_training_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        train_data = data.get_dataset(self.download_directory, train=True)
+        return DataLoader(train_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def build_validation_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        validation_data = data.get_dataset(self.download_directory, train=False)
+        return DataLoader(validation_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def train_batch(
+        self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs, _ = batch
+
+        self.generator.requires_grad_(True)
+        self.discriminator.requires_grad_(False)
+
+        # sample noise and match gpu device (or keep as cpu)
+        z = torch.randn(imgs.shape[0], self.context.get_hparam("latent_dim"))
+        z = self.context._to_device(z)
+
+        # generate images and log sampled images to Tensorboard
+        generated_imgs = self.generator(z)
+        sample_imgs = generated_imgs[:6]
+        grid = torchvision.utils.make_grid(sample_imgs)
+        self.logger.writer.add_image(f'generated_images_epoch_{epoch_idx}', grid, batch_idx)
+
+
+        # train generator
+        valid = torch.ones(imgs.size(0), 1)
+        valid = self.context._to_device(valid)
```

Comment:
Explain why the `to_device()` call is necessary, since it's really not intuitive why this tensor is special, from the perspective of somebody who is new to our system.

Reply:
Add comments to remind users since any created tensor should be called `to_device`